File: js_files/003lengthOfLongestSubstring.js
const lengthOfLongestSubstring = function(s) {
    const occ = new Set();
    const n =s.length;
    let rk = -1, ans = 0;
    for (let i = 0; i < n; ++i) {
        if (i != 0) {
            occ.delete(s.charAt(i - 1));
        }
        while (rk + 1 < n && !occ.has(s.charAt(rk + 1))) {
            occ.add(s.charAt(rk + 1));
            ++rk;
        }
        ans = Math.max(ans, rk - i + 1);
    }
    return ans;
};

// 测试函数
function testLengthOfLongestSubstring() {
    const testCases = [
        { input: "abcabcbb", expected: 3 }, // 最长无重复子串是 "abc"
        { input: "bbbbb", expected: 1 },    // 最长无重复子串是 "b"
        { input: "pwwkew", expected: 3 },   // 最长无重复子串是 "wke"
        { input: "", expected: 0 },         // 空字符串，返回 0
        { input: " ", expected: 1 },        // 只有一个空格，返回 1
        { input: "abcdef", expected: 6 },   // 全部都是不同字符，返回 6
        { input: "abba", expected: 2 },     // 最长无重复子串是 "ab" 或 "ba"
        { input: "dvdf", expected: 3 },     // 最长无重复子串是 "vdf"
        { input: "tmmzuxt", expected: 5 }   // 最长无重复子串是 "mzuxt"
    ];

    testCases.forEach(({ input, expected }, index) => {
        const result = lengthOfLongestSubstring(input);
        console.log(`Test ${index + 1}: Input = "${input}", Output = ${result}, Expected = ${expected}`);
        console.assert(result === expected, `Test ${index + 1} failed: expected ${expected}, got ${result}`);
    });

    console.log("All tests completed.");
}

// 运行测试
testLengthOfLongestSubstring();

File: go_files/003lengthOfLongestSubstring.go
package main

import "fmt"

func lengthOfLongestSubstring(s string) int {
	//哈希结合,记录每个字符是否出现过
	m := map[rune]int{}
	n := len(s)
	//右指针,初始值为-1,相当于在字符串的左边界的左侧,还没开始移动
	rk, ans := -1, 0
	for i := 0; i < n; i++ {
		if i != 0 {
			//左指针向右移动一格,移除一个字符
			delete(m, rune(s[i-1]))
		}
		for rk+1 < n && m[rune(s[rk+1])] == 0 {
			//不断地移动右指针
			m[rune(s[rk+1])]++
			rk++
		}
		//第1到rk个字符是一个极长地无重复字符串
		ans = max(ans, rk-i+1)
	}
	return ans
}

func max(x, y int) int {
	if x < y {
		return y
	}
	return x
}

func main() {
	testString := "testString"
	fmt.Println(lengthOfLongestSubstring(testString))
}
